mantidproject/mantid
# Reflectometry GUI Restructuring 10: Autoprocess

~~***PR #24923 should be merged first.***~~

**Description of work.**

Ongoing work to integrate the `JobTreeView` into the `ISIS Reflectometry` GUI. This PR is being merged to a feature branch because not all functionality is implemented yet.

This PR adds auto-processing of runs from an experiment. This involves processing all existing runs for an experiment and polling for new runs. Polling is done via a `QBasicTimer` which is owned by the `RunsView`. A new class `CatalogRunNotifier` wraps the `RunsView`. It has methods to start/stop polling and subscribes to the view's timer notifications which are then passed through to the `RunsPresenter` via the `IRunNotifier` interface. Currently the logic to search for new runs is still in the `RunsPresenter` but the aim is to move this to the `CatalogRunNotifier` in a future PR.

This PR also adds improved transfer of runs from the search results box - duplicate runs are now excluded, and runs with the same angle will now always be added to the same row in the table (previously they would be added to separate rows if they were transferred separately).

**To test:**

- Log into the catalog from the `Catalog` menu
- Open the `ISIS Reflectometry` interface.
- Enter the instrument as `INTER` and the search string as `1120015`
- Click `Autoprocess`
- The table should be populated with runs from that experiment and should start processing them.
- Check that you can pause autoreduction and re-start it.
- Try playing around with manually processing and pausing individual rows, deleting and adding rows etc. 
- Refer to the [documentation](https://docs.mantidproject.org/nightly/interfaces/ISIS%20Reflectometry.html#autoprocess) for guidance on the expected behaviour.
- Clear the table and try transferring some runs manually with the `Transfer` button. They should be placed in the correct group whether transferred individually or all together (e.g. in this screenshot, multiple runs are added to the same row, even if transferred separately):
![transfer_runs](https://user-images.githubusercontent.com/12895056/57936084-18b14900-78bb-11e9-8f38-2edc6d90f547.png)
- Check that when transferring rows, they are always sorted by increasing angle within the group regardless of the order they were added. This is important because otherwise stitching fails. Note however that groups are not sorted so they will appear in the order they were added.
- If possible, leave autoprocessing running on a live experiment and check that when new runs appear in the archive that they are found and processed in the reflectometry GUI. (I use [JournalViewer](https://www.projectaten.com/jv) to check for the current experiment and when new runs come in). Note that the most recent run may not appear at the bottom of the table if it is in a group that already exists - you can use the filter to easily find it.

Refs #23027 

*This does not require release notes* because it re-implements functionality from the current interface.

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
